File: lc/coreprep/implementation/patterns_reference_implementation/tree_dfs/SumOfPathNumbers.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(n)
public class SumOfPathNumbers {
  static class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
      this.left = null;
      this.right = null;
    }
  }

  private static TreeNode buildTree(String[] nodes) {
    return buildTree(nodes, 0);
  }

  private static TreeNode buildTree(String[] nodes, int currIdx) {
    if (currIdx >= nodes.length || nodes[currIdx].equalsIgnoreCase("null")) {
      return null;
    }

    TreeNode root = new TreeNode(Integer.parseInt(nodes[currIdx]));
    root.left = buildTree(nodes, 2 * currIdx + 1);
    root.right = buildTree(nodes, 2 * currIdx + 2);

    return root;
  }

  private static int sumOfPathNumbers(TreeNode root) {
    return sumOfPathNumbers(root, 0);
  }

  private static int sumOfPathNumbers(TreeNode root, int sum) {
    if (root == null) {
      return 0;
    }

    int currSum = 10 * sum + root.val;

    if (root.left == null && root.right == null) {
      return currSum;
    }

    return sumOfPathNumbers(root.left, currSum) +
        sumOfPathNumbers(root.right, currSum);
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- > 0) {
        String[] nodes = in.nextLine().trim().split(" ");
        TreeNode root = buildTree(nodes);
        System.out.println(sumOfPathNumbers(root));
      }
    }
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/tree_dfs/sum_of_path_numbers.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

TreeNode *build_tree(const vector<string> &nodes, int curr_idx) {
  if (curr_idx >= nodes.size() || nodes[curr_idx] == "null") {
    return nullptr;
  }

  TreeNode *root = new TreeNode(stoi(nodes[curr_idx]));
  root->left = build_tree(nodes, 2 * curr_idx + 1);
  root->right = build_tree(nodes, 2 * curr_idx + 2);

  return root;
}

TreeNode *build_tree(const vector<string> &nodes) {
  return build_tree(nodes, 0);
}

int sum_of_path_numbers(TreeNode *root, int sum) {
  if (root == nullptr) {
    return 0;
  }

  int curr_sum = 10 * sum + root->val;

  if (root->left == nullptr && root->right == nullptr) {
    return curr_sum;
  }

  return sum_of_path_numbers(root->left, curr_sum) +
         sum_of_path_numbers(root->right, curr_sum);
}

int sum_of_path_numbers(TreeNode *root) { return sum_of_path_numbers(root, 0); }

// O(n) / O(n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt;
  string s;

  cin >> tt;
  cin.ignore(1);

  while (tt--) {
    getline(cin, s);

    string tmp;
    stringstream ss(s);
    vector<string> nodes;

    while (getline(ss, tmp, ' ')) {
      nodes.push_back(tmp);
    }

    TreeNode *root = build_tree(nodes);
    cout << sum_of_path_numbers(root) << "\n";
  }

  return 0;
}